gem/lib/textmate_grammar/grammar.rb

```diff
@@ -0,0 +1,560 @@
+            end
+        end
+
+        output[:repository] = repo.transform_values { |value| to_tag.call(value) }
+
+        output[:patterns] = output[:repository][:$initial_context]
+        output[:patterns] ||= []
+        output[:patterns] = output[:patterns]["patterns"] if output[:patterns].is_a? Hash
+        output[:repository].delete(:$initial_context)
+
+        output[:version] = auto_version
+        output.merge!(@keys) { |_key, old, _new| old }
+
+        @@transforms.each { |transform| output = transform.post_transform(output) }
+
+        @@linters.each do |linter|
+            raise "linting failed, see above error" unless linter.post_lint(output)
+        end
+
+        output
+    end
+
+    #
+    # Save the grammar to a path
+    #
+    # @param [Hash] options options to save_to
+    # @option options :inherit_or_embedded (:embedded) see {#generate}
+    # @option options :generate_tags [Boolean] (true) generate a list of all +:tag_as+s
+    # @option options :dir [String] the location to generate the files
+    # @option options :tag_dir [String] (File.join(options[:dir],"language_tags")) the
+    #   directory to generate language tags in
+    # @option options :syntax_dir [String] (File.join(options[:dir],"syntaxes")) the
+    #   directory to generate the syntax file in
+    # @option options :syntax_format [:json,:vscode,:plist,:textmate,:tm_language,:xml]
+    #   (:json) The format to generate the syntax file in
+    # @option options :syntax_name [String] ("#{@name}.tmLanguage") the name of the syntax
+    #   file to generate without the extension
+    #
+    # @note all keys except :dir is optional
+    # @note :dir is optional if both :tag_dir and :syntax_dir are specified
+    # @note currently :vscode is an alias for :json
+    # @note currently :textmate, :tm_language, and :xml are aliases for :plist
+    # @note later the aliased :syntax_type choices may enable compatibility features
+    #
+    # @return [void] nothing
+    #
+    def save_to(options)
+        options[:dir] ||= "."
+        default = {
+            inherit_or_embedded: :embedded,
+            generate_tags: true,
+            syntax_format: :json,
+            syntax_name: "#{@scope_name.split('.').drop(1).join('.')}.tmLanguage",
+            syntax_dir: File.join(options[:dir], "syntaxes"),
+            tag_dir: File.join(options[:dir], "language_tags"),
+        }
+
+        options = default.merge(options)
+        output = generate(options[:inherit_or_embedded])
+
+        if [:json, :vscode].include? options[:syntax_format]
+            file_name = File.join(
+                options[:syntax_dir],
+                "#{options[:syntax_name]}.json",
+            )
+            out_file = File.open(file_name, "w")
+            out_file.write(JSON.pretty_generate(output))
+            out_file.close
+        elsif [:plist, :textmate, :tm_language, :xml].include? options[:syntax_format]
```

Comment:
Isn't the textmate format different from XML? I thought it looked kind of like json, but with `()` instead of `{}` and `=` instead of `:`

Reply:
That's old style property list format which textmate shows for convenience, but the file is xml style plist.